### ISSUE TITLE
[meta] Small clean-ups welcome?

### ISSUE THREAD

**Issue by stylewarning** [fricas/fricas#39]
Hello, I hope this is an OK place to ask a meta-question about the project. I'm Robert and I work in quantum computing, and I am also a long-time, professional Common Lisp programmer (i.e., I use it for my job).

First, thanks to the FriCAS developers for keeping the project alive and well. I hope that one day FriCAS picks up even more steam and is continued to be built upon and developed. Computer algebra is hard, and building on existing systems seems like our best bet for building a better CAS.

I've been interested in contributing back to FriCAS for many years, and I would like to start now. I have a few newbie questions.

1. Is it OK to use the issue tracker for questions, or do the maintainers highly prefer using a mailing list?
2. Does this project accept non-breaking Lisp code cleanup? I'm afraid I can't contribute within the Spad language itself yet because I don't sufficiently understand it, but when I see files like [this](https://github.com/fricas/fricas/blob/master/src/lisp/primitives.lisp), I wonder if we can use modern Common Lisp features like `inline`d functions instead of macros everywhere.
3. Does this project support deleting dead code or deprecated platforms? I don't think supporting GCL or Poplog is worth it; there are better, more portable, more performant, and more maintained Common Lisp systems. Deleting code is always scary, but it's better than letting it secretly bitrot.
4. How does this project handle dependencies? Would project maintainers accept FriCAS being built through the idiomatic ASDF? If so, that would make depending on third-party Lisp code a *lot* easier.

Thank you!

**Comment by oldk1331** [comment]
1. It's OK to use the GitHub issues (for users to raise questions), but I think most FriCAS developers would prefer the mailing list (for development discussion).
2. It would be nice to have Lisp code cleanup, we can review it.
3. I think we are open for discussion.
4. This project almost has zero dependencies.  Zero Common Lisp dependency.  Very very few C dependency.  Having ASDF would be a nice plus.  https://github.com/nilqed/fricas0 would be a good starting point for you.

**Comment by hebisch** [comment]
On Fri, Mar 05, 2021 at 04:02:03PM -0800, Robert Smith wrote:
> Hello, I hope this is an OK place to ask a meta-question about the project. I'm Robert and I work in quantum computing, and I am also a long-time, professional Common Lisp programmer (i.e., I use it for my job).
> 
> First, thanks to the FriCAS developers for keeping the project alive and well. I hope that one day FriCAS picks up even more steam and is continued to be built upon and developed. Computer algebra is hard, and building on existing systems seems like our best bet for building a better CAS.
> 
> I've been interested in contributing back to FriCAS for many years, and I would like to start now. I have a few newbie questions.
> 
> 1. Is it OK to use the issue tracker for questions, or do the maintainers highly prefer using a mailing list?

I prefer tracker for specific issues and mailing list for discussion.
One problem of issue tracker (and webby interfaces in general) is
that it tends to mangle plain text.  Advantage of issue tracker
is that messages are associated to specific issue.  However,
discussions usually have topic drift and advantage of
"specific issue" is lost if used for discussion (and disadvantages
remain).

> 2. Does this project accept non-breaking Lisp code cleanup? I'm afraid I can't contribute within the Spad language itself yet because I don't sufficiently understand it, but when I see files like [this](https://github.com/fricas/fricas/blob/master/src/lisp/primitives.lisp), I wonder if we can use modern Common Lisp features like `inline`d functions instead of macros everywhere.

Actual cleanups are welcome.  However, note that FriCAS have reasons
for coding practices that are different from usual Lisp good style.
One aspect is interaction between four languages: C, Lisp, Boot and
Spad.  It is _much_ easier to work with code if "the same" object
in different languages have exactly the same name, and this in
many cases implies using upper case Lisp names.  Second,
some part in FriCAS are computationaly intensive, and we want
as fast object code as we can get.  Third, Spad compiler
have some understanding of lists and vectors, but can not
use structs or classes.  So Lisp constructs used from Spad
are necessarly limited to rather small core.  Fourth, we
distrubute also binaries.  New systems have some support
for running old binaries, while frequently it is not possible
to run new binaries on old systems.  So to cover many
system our binaries use only old featurs.

On constructive side, there are various possible contribution
at Lisp level.  Our Lisp code has long history, there were
rather intesive recent cleanups, but surly there is potential
for more cleanups.  Code in "primitives.lisp" is based on
profiling to get good performance, but Lisp systems
evolve and some new idioms potentially may give better
performance.  At Spad level FriCAS currently provides
tracing, but hooks into Lisp (like call stack walker)
could improve debugging.

> 3. Does this project support deleting dead code or deprecated platforms?

Deleting dead code: sure.  A lot of dead code were removed, but
still some remains to be found.

> I don't think supporting GCL or Poplog is worth it; there are better,
> more portable, more performant, and more maintained Common Lisp systems.

Well, concerning Poplog, I had significant hope related to it.
There are problems, but I did not entirely gave up.  Related code
in FriCAS is tiny, so hopefully should not bother you too much.

GCL is different case.  Some parts of Lisp code are written in
strange way because to was the only way that worked in GCL.
So, there would be some gain from dropping GCL compatibility.
OTOH on Debian GCL is well supported, AFAIK it was (and possibly is)
the only Lisp that worked on all Debian architectures.  Concerning
performance, GCL gives us somewhat better performance than ECL.
Also, Camm Maguire in the past and quite recently worked to
make sure that FriCAS works with GCL.  So I do not think that
we should to drop GCL suppot.  It makes sense to restrict
GCL support to ANSI flavour, that is drop support for
"traditional" non-ANSI GCL.

> Deleting code is always scary, but it's better than letting it secretly bitrot.

FYI, I test FriCAS with various Lisps and except for Poplog (which
would need more work for "production" support) the officially
supported ones worked.  Newest non-ANSI GCL does not work with
FriCAS (older, confusingly having the same version number works).
Untested is Lispworks and CMUCL.  They are more recent than
other ports and presumably worked when contributed.  My view
is that if somebody contributed code then presumably there is
some interest in running FriCAS on those platforms.  And
fixing code that used to work is easier than starting from
scratch.  To put it differently, in general I try to avoid
breaking code.  But beside such general care I put no special
effort into Lispworks and CMUCL port.  But I also do not
"depreciate" such platforms: support is as good as users
of those platforms contribute.

> 4. How does this project handle dependencies? Would project maintainers accept FriCAS being built through the idiomatic ASDF? If so, that would make depending on third-party Lisp code a *lot* easier.

Well, there was probably about 1 man-year of developement effort
spent on current FriCAS build system (most of that in AXIOM era,
before FriCAS was established as independent project).  The result
is satisfactory in the sense that it is now low-maintanence
well-working system.  Before work on current build system AXIOM
went trough several different build systems, having various
problems, some were quite buggy.  So I am very sceptical about
changes to build system, mistakes there can be large time sink.
Concerning third-party Lisp code, that needs ASDF, I think that
there should be a way to accomodate ASDF inside current build
system.  That should be _much_ easier than trying to use ASDF
for other tasks.

-- 
                              Waldek Hebisch
